Salvage mob restrictions

20kdc opened the pull request:
## About the PR

Written due to popular administrative demand, this:

+ Kills salvage-spawned mobs when the salvage despawns
+ Makes salvage-spawned mobs eligible for deletion when the salvage despawns even if player-occupied
+ Tweaks salvage mob ghost role stuff so that they don't have any of this weird business about "well if you're NOT part of a salvage wreck then go wreak havoc".

This should hopefully keep them from causing too much trouble that the administration decide to flat out remove them.

**Changelog**

:cl:
- tweak: Salvage mobs will now die if they're off the salvage wreck when it despawns.

retequizzle commented:
I actually really like this. It almost fully mirrors what I would want changed with the ghost role to keep them playable without invalidating them entirely. 

If people are that dead set on being space fish to where they'll spend the whole round doing it, then the role is too potent and needs to be looked at.

20kdc commented:
Rebased and modified a bit - master did an upstream change that was a bit more strict with which mobs got Puller, and said changes were a subset of my changes, so it auto-applied just fine but my changes still ended up re-adding puller (because I let non-space mobs still have it).
So basically this PR no longer touches the presence or lack thereof of Puller on mobs.

mirrorcult commented on Content.Server/Salvage/SalvageMobRestrictionsSystem.cs:
```diff
@@ -0,0 +1,74 @@
+    {
+        foreach (EntityUid target in component.MobsToKill)
+        {
+            if (TryComp(target, out DamageableComponent? dc))
+            {
+                _damageableSystem.SetAllDamage(dc, 200);
+            }
```
it should check if they have a body first to try gibbing them

20kdc replied:
The involved mobs here are simplemobs and can't be gibbed.
I guess I could add the check in case of a future non-simplemob.

20kdc replied:
Done as of 6ee4019

mirrorcult reviewed (approved):
not the happiest about this since i think it'd be more interesting if they weren't restricted, but we can't really trust our playerbase with it, sadly

retequizzle commented:
on the upside, having it kill salvage mobs on despawn means we're one step closer to having it gib player mobs on despawn too

20kdc commented:
> on the upside, having it kill salvage mobs on despawn means we're one step closer to having it gib player mobs on despawn too

actually it used to delete players that stuck around
then someone changed it

retequizzle commented:
> > on the upside, having it kill salvage mobs on despawn means we're one step closer to having it gib player mobs on despawn too
> 
> actually it used to delete players that stuck around then someone changed it

i know it deletes player corpses, if that got removed that actually kind of sucks. the whole point of salvage is to be off it when it goes or you risk going with it.

oh well.

retequizzle commented:
> Weelllll
> 
> I think the logic was along the lines of that it deletes player corpses if and only if the player is not inhabiting the corpse at the time

yeah that falls in line with what i've seen/experienced as well. though to be honest i haven't tested if it occurs when the ghost is still in the body, that might be worth looking into.

20kdc commented:
> > Weelllll
> > I think the logic was along the lines of that it deletes player corpses if and only if the player is not inhabiting the corpse at the time
> 
> yeah that falls in line with what i've seen/experienced as well. though to be honest i haven't tested if it occurs when the ghost is still in the body, that might be worth looking into.

unless I read the logic incorrectly, if the ghost is still in the body it's safe from deletion
